lib/autoproj/cli/envsh.rb

```diff
@@ -1,19 +1,147 @@
+require 'rbconfig'
 require 'autoproj/cli/inspection_tool'
 module Autoproj
     module CLI
         class Envsh < InspectionTool
-            def validate_options(_unused, options = Hash.new)
-                _, options = super(_unused, options)
+            attr_predicate :watch?
+            attr_reader :options
+            attr_reader :source_packages
+            attr_reader :notifier
+            attr_reader :watchers
+            def validate_options(unused, options = {})
+                _, options = super(unused, options)
+                @watchers = []
+                @watch = options[:watch]
+                @options = options
                 options
             end
 
-            def run(**options)
+            def update_workspace(restart = false)
+                exec('autoproj', 'envsh', '--watch') if restart
+
                 initialize_and_load
-                shell_helpers = options.fetch(:shell_helpers, ws.config.shell_helpers?)
-                finalize_setup(Array.new)
+                shell_helpers = options.fetch(:shell_helpers,
+                                              ws.config.shell_helpers?)
+                @source_packages, = finalize_setup([])
                 export_env_sh(shell_helpers: shell_helpers)
             end
+
+            def callback
+                puts 'Workspace changed...'
+                stop_watchers
+                update_workspace(true)
+            end
+
+            def create_file_watcher(file)
+                watchers << notifier.watch(file, :modify) do |e|
+                    delete_watcher(e.watcher) if e.flags.include? :ignored
+                    callback
+                end
+            end
+
+            def create_dir_watcher(dir, *files, recursive: false)
+                opt_args = %i[move create delete]
```

Comment:
I would prefer having the static arguments directly in the call for clarity:

`watchers << notify.watch(dir, :move, :create, :delete, *opt_args) ...`